javalib/src/main/scala/java/util/Formatter.scala

```diff
@@ -1425,4 +1359,242 @@ object Formatter {
     def toUpperCase(str: String): String = str.toUpperCase(actualLocale)
   }
 
+  /* ParserStateMachine ported from Apache Harmony,
+   * Used instead of regexp due to performance issues
+   */
+  private class ParserStateMachine(format: CharBuffer) {
+
+    import ParserStateMachine._
+
+    def nextFormatToken(): FormatToken = {
+      var currentChar = FormatToken.Unset.asInstanceOf[Char]
```

Comment:
```suggestion
      var currentChar = FormatToken.Unset.toChar
```
Or probably better, define a separate `val UnsetChar = '\u0000'`.